Example orderbook subscribe sol.

papadpickle opened the pull request:
this example subscribes to SOL perp market on-chain events for bids, asks and fill.
following info is logged at each update of each orderbook as well as fill event:
* latest trade
* bid-ask values
* midprice
* spread in bps
* market depth at +/-2%

note: in case no fill event received before an orderbook update, the latest trade will be empty. maybe there is a way to get the last traded price(mango-bowl recent trades maybe?), in that case we can add that in future. 

snippet:
![image](https://user-images.githubusercontent.com/99397192/155863696-7c9a58db-77af-4afc-9a7b-63b8089548fe.png)

additions:
* a subscription::wssSubscriber class, which can be/is reused to subscribe to given uri. callbacks can be registered by its clients to handle the incoming wss messages.
* a subscription::bookSide class, which uses wssSubscriber to subscribe to bids and asks and offers its clients interfaces to get best prices and volume. additionaly it can notify the clients on each update via a callback.
* an orderbook::book class which contains two bookSide instances representing bids and asks and offers its clients an interface to get the level1 info.
* a subscription::trades class, which uses wssSubscriber to subscribe to market event queue and offers its clients interfaces to get last traded price. additionaly it can notify the clients on each update via a callback.
* a orderbook::level1 and orderbook::order structs

minor improvements:
* const correctness at solana api
* added todo to use scoped enums at existing mango code, didnt change it here as this might requires changes in other examples as well.

mschneider commented on include/subscriptions/wssSubscriber.hpp:
```diff
@@ -0,0 +1,100 @@
+
+      websocketpp::lib::error_code ec;
+      ws_client::connection_ptr con = c.get_connection(
+          "wss://mango.rpcpool.com/946ef7337da3f5b8d3e4a34e7f88", ec);
```
add a new constant to mango.hpp

papadpickle replied:
yea, MAINNET.endpoint should work.

mschneider commented:
the order book parsing code that i sent you wasn't complete yet, it doesn't skip orders that are not "in force" anymore:

https://github.com/blockworks-foundation/mango-v3/compare/v3.3.5...v3.4.1#diff-22492bdca835846d3f1d6f98c78589aedab06775fa130eae66fe2f6c87a89a08R180

mschneider commented:
the order book parsing code that i sent you wasn't complete yet, it doesn't skip orders that are not "in force" anymore:

https://github.com/blockworks-foundation/mango-v3/compare/v3.3.5...v3.4.1#diff-22492bdca835846d3f1d6f98c78589aedab06775fa130eae66fe2f6c87a89a08R180

mschneider commented:
``` 
[2022-03-01 20:24:16.874] [info] ============Update============
[2022-03-01 20:24:16.874] [info] Latest trade: 9880
[2022-03-01 20:24:16.874] [info] Bid-Ask 9879-9880
[2022-03-01 20:24:16.874] [info] MidPrice: 9879.5
[2022-03-01 20:24:16.874] [info] Spread: 1.01 bps
[2022-03-01 20:24:16.874] [info] Market depth -2%: 1458616
[2022-03-01 20:24:16.874] [info] Market depth +2%: 950953
``` 

output should probably display ui units instead of native units, i'll open a separate ticket for this

mschneider commented on include/subscriptions/bookSide.hpp:
```diff
@@ -0,0 +1,121 @@
+    if (!newOrders.empty()) {
+      {
+        std::scoped_lock lock(ordersMtx);
+        orders = std::move(newOrders);
```
data should already be sorted by price as the order book tree datastructure is sorted by price

papadpickle replied:
ok. will have to make sure though that the bidside and askside sorting is from midpoint to either ends.

mschneider commented on include/subscriptions/bookSide.hpp:
```diff
@@ -0,0 +1,121 @@
+            leafNode->timestamp + leafNode->timeInForce < nowUnix;
+        if (isValid) {
+          newOrders.emplace_back((uint64_t)(leafNode->key >> 64),
+                                 leafNode->quantity);
```
i have a few issues with the way the code is organized:
1) parsing the order tree should be independent from how the data is transported (rpc / ws), the code should hence not live under subscriptions
2) same is true for getVolume
3) each LeafNode actually contains a lot of valuable information that is discarded here in favor of the very incomplete order struct, this makes this code very hard to re-use (we have an order type in most sdks it usually contains more data then the LeafNode, ui values for instance)

papadpickle replied:
will ping you on discord to catch up here.

alexeyneu replied:
where

mschneider commented:
tried it locally and the code works well for a few minutes but then one by one the different sockets disconnect. once a single socket disconnects, aggregated data can be invalid (e.g. mid price, after one side of the book stopped updating)

mschneider commented on include/mango_v3.hpp:
```diff
@@ -135,8 +138,10 @@ struct EventQueueHeader {
   uint64_t seqNum;
 };
 
+// todo: change to scoped enum class
```
sounds good, want to open a new issue?

papadpickle replied:
yes

papadpickle commented:
added changes to stop logging and exit in case of problem at any of the wss connections.
![image](https://user-images.githubusercontent.com/99397192/156380413-80fc7629-e38b-46fa-af15-9e9fe11778a9.png)

papadpickle commented:
re your comment https://github.com/mschneider/solcpp/pull/19#discussion_r817093630: 
1. parsing of the messages is moved to the classes(BookSide and Trade in mango_v3.hpp) out of subscriptions. 
2. handled at point 1
3. this is first iteration of parsing the incoming messages and extracted only necessary info. We can fill up/edit the parsing and data struct as we go was my idea. You can create an issue here which can be handled later. 
4. I have refactored the code by adding AccountSubscriber<T> class, removed unnecessary subscriber classes and replacing critical sections with shared_ptr. 

I have one point about the structure of the code overall:
we can refactor the monolithic mango_v3.hpp into individual hpp and cpp files to speed up dev work. Now a single line change at those interfaces forces the recompilation of everything. Since we are anyway creating a static lib, we may as well go with seperate files instead of having one huge hpp.
Unless of course you decide to make it a header only lib.

mschneider commented on include/subscriptions/orderbook.hpp:
```diff
@@ -0,0 +1,67 @@
+  }
+
+ private:
+  std::shared_ptr<L1Orderbook> level1 = std::make_shared<L1Orderbook>();
```
nullptr makes sense here to differentiate between waiting for first data update vs empty book

papadpickle replied:
makes sense. will change it.

mschneider commented on include/mango_v3.hpp:
```diff
@@ -197,6 +203,247 @@ struct EventQueue {
+
+ private:
+  uint64_t lastSeqNum = INT_MAX;
+  std::shared_ptr<uint64_t> latestTrade = std::make_shared<uint64_t>(0);
```
why not just a uint64_t?

mschneider commented on include/mango_v3.hpp:
```diff
@@ -197,6 +203,247 @@ struct EventQueue {
+      if ((*iter).tag == NodeType::LeafNode) {
+        const auto leafNode =
+            reinterpret_cast<const struct LeafNode *>(&(*iter));
+        const auto now = std::chrono::system_clock::now();
```
you dont want to recalculate the current time inside the loop

papadpickle replied:
https://github.com/mschneider/solcpp/blob/9deb4c90abec347db7cdf288c7cc2d603d65b7aa/examples/ticker.cpp#L36
mango gravy from the initial example lol. 
pls create a seperate bug/issue for this.

mschneider commented on include/mango_v3.hpp:
```diff
@@ -197,6 +203,247 @@ struct EventQueue {
+
+ private:
+  uint64_t lastSeqNum = INT_MAX;
+  std::shared_ptr<uint64_t> latestTrade = std::make_shared<uint64_t>(0);
```
why not just an uint64_t, trade at price 0 is impossible after all

papadpickle replied:
will need a mtx/atomic otherwise. same reason why we replaced all mtx'es with shared ptrs. 
event based synchronization between getLastTrade and handleMsg.
can change to nullptr though so client of getLastTrade could parse the validity of returned value.

mschneider commented on include/mango_v3.hpp:
```diff
@@ -197,6 +203,247 @@ struct EventQueue {
+ public:
+  BookSide(Side side) : side(side) {}
+
+  bool handleMsg(const nlohmann::json &msg) {
```
this should really live in the websocket subscription code, rather than the account type. in fact the code should work the same for all types of account subscriptions.

papadpickle replied:
the first four lines to check json "result" or the whole of msg handling? 
first four lines could be moved one layer below to wss subscription code in case we know that each msg has the "result" object. 
the other parsing code must be handled by the account itself.

papadpickle commented on include/mango_v3.hpp:
```diff
@@ -197,6 +203,247 @@ struct EventQueue {
+    // all other messages are event queue updates
+    const std::string method = msg["method"];
+    const int subscription = msg["params"]["subscription"];
+    const int slot = msg["params"]["result"]["context"]["slot"];
```
well the gravy for mango related code was taken from the original examples from the first commit.
https://github.com/mschneider/solcpp/blob/dd58f26432402c9dd4721811a38aab5514628c29/examples/accountSubscribe.cpp#L69

you can create a bug/issue for this. out of scope for this imo.

papadpickle commented:
updated the PR with following changes: 
- https://github.com/mschneider/solcpp/pull/19#discussion_r840560642 -> decoding in the account subscriber itself.
- handling bookTimeDelay to filter expired orders
- using FillEvent struct for trades.
- using LeafNode struct as order struct
- minor improvements

mschneider commented on include/mango_v3.hpp:
```diff
@@ -259,6 +266,264 @@ struct EventQueue {
+                               std::to_string(decoded.size()) + " expected " +
+                               std::to_string(sizeof(BookSideRaw)));
+    }
+    memcpy(&(*raw), decoded.data(), sizeof(BookSideRaw));
```
needs critical section from before memcpy to end of iterator lifetime.
concurrent update calls else could cause memory corruption.
another option would be to not update the same buffer and to swap references to a new buffer, so that iterator can still iterate over the "old buffer" and de-allocate it on iterator destruction.

mschneider commented on include/mango_v3.hpp:
```diff
@@ -259,6 +266,264 @@ struct EventQueue {
+  uint64_t getVolume(uint64_t price) const {
+    Op operation;
+    uint64_t volume = 0;
+    for (auto&& order : *orders) {
```
orders needs to be dereferenced once before entering loop

godmodegalactus commented on include/subscriptions/wssSubscriber.hpp:
```diff
@@ -0,0 +1,115 @@
+
+    try {
+      websocketpp::lib::error_code ec;
+      ws_client::connection_ptr con = c.get_connection(MAINNET.endpoint, ec);
```
`MAINNET.endpoint` should not be statically passed but we should create an endpoint argument in the constructor.

alexeyneu replied:
again:
